python/lsst/ip/diffim/imageMapReduce.py

```diff
@@ -0,0 +1,575 @@
+from __future__ import absolute_import, division, print_function
+from future import standard_library
+standard_library.install_aliases()
+#
+# LSST Data Management System
+# Copyright 2016 AURA/LSST.
+#
+# This product includes software developed by the
+# LSST Project (http://www.lsst.org/).
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the LSST License Statement and
+# the GNU General Public License along with this program.  If not,
+# see <https://www.lsstcorp.org/LegalNotices/>.
+#
+
+import numpy as np
+import abc
+from future.utils import with_metaclass
+
+import lsst.afw.image as afwImage
+import lsst.afw.math as afwMath
+import lsst.afw.geom as afwGeom
+import lsst.pex.config as pexConfig
+import lsst.pipe.base as pipeBase
+
+__all__ = ("ImageMapReduceTask", "ImageMapReduceConfig",
+           "ImageMapperSubtask", "ImageMapperSubtaskConfig",
+           "ImageReducerSubtask", "ImageReducerSubtaskConfig")
+
+
+class ImageMapperSubtaskConfig(pexConfig.Config):
+    """Configuration parameters for the ImageMapperSubtask
+    """
+    pass
+
+
+class ImageMapperSubtask(with_metaclass(abc.ABCMeta, pipeBase.Task)):
+    """Abstract base class for any task that is to be
+    used as `ImageMapReduceConfig.mapperSubtask`
+    """
+    ConfigClass = ImageMapperSubtaskConfig
+    _DefaultName = "ip_diffim_ImageMapperSubtask"
+
+    def run(self, subExp, expandedSubExp, fullBBox, **kwargs):
+        """Perform operation on given sub-exposure.
+
+        To be implemented by subclasses.
+
+        Parameters
+        ----------
+        subExp : afw.Exposure
+            the sub-exposure upon which to operate
+        expandedSubExp : afw.Exposure
+            the expanded sub-exposure upon which to operate
+        fullBBox : afwGeom.BoundingBox
+            the bounding box of the original exposure
+        kwargs :
+            additional keyword arguments
+
+        Returns
+        -------
+        anything, including an a `afw.Exposure`
+        """
+        pass
+
+
+class ImageReducerSubtaskConfig(pexConfig.Config):
+    """Configuration parameters for the ImageReducerSubtask
+    """
+    reduceOperation = pexConfig.ChoiceField(
+        dtype=str,
+        doc="""Operation to use for reducing subimages into new image.""",
+        default="copy",
```

Comment:
I guess that's ok? I just worry that someone would change the other to have overlaps and not notice this comment.